components/sidebar/sidebar_channel/sidebar_channel_menu/sidebar_channel_menu.tsx

```diff
@@ -1,284 +1,502 @@
 
-export class SidebarChannelMenu extends React.PureComponent<Props, State> {
-    isLeaving: boolean;
+const SidebarChannelMenu = (props: Props): JSX.Element => {
+    const {channel, channelLink, location} = props;
+
+    const buttonReference = useRef<HTMLButtonElement>(null);
+    const menuItemReference = useRef(null);
+    const [isMenuVisible, setMenuVisible] = useState(false);
+    const [isSubMenuVisible, setSubMenuVisible] = useState(false);
+    const intl = useIntl();
+    const dispatch = useDispatch<DispatchFunc>();
+
+    //****************** Props ******************//
+    const member = useSelector(
+        (state: GlobalState) => getMyChannelMemberships(state)[channel.id],
+    );
+
+    const isUnread = useSelector((state: GlobalState) =>
+        getDisplayedChannels(state),
+    );
+
+    const currentUserId = useSelector((state: GlobalState) =>
+        getCurrentUserId(state),
+    );
+    const isFavorite = useSelector((state: GlobalState) =>
+        isFavoriteChannel(state, channel.id),
+    );
+    const isMuted = useSelector(() => isChannelMuted(member));
+    const displayedChannels = useSelector((state: GlobalState) =>
+        getDisplayedChannels(state),
+    );
+    const multiSelectedChannelIds = useSelector(
+        (state: GlobalState) =>
+            state.views.channelSidebar.multiSelectedChannelIds,
+    );
+    const currentTeam = useSelector((state: GlobalState) =>
+        getCurrentTeam(state),
+    );
+
+    //****************** Menu Item Actions ******************//
+
+    const handleLeavePublicChannel = () => {
+        dispatch(leaveChannel(channel.id));
+        trackEvent('ui', 'ui_public_channel_x_button_clicked');
+    };
 
-    constructor(props: Props) {
-        super(props);
+    const handleLeavePrivateChannel = () => {
+        dispatch(
+            openModal({
+                modalId: ModalIdentifiers.LEAVE_PRIVATE_CHANNEL_MODAL,
+                dialogType: LeavePrivateChannelModal,
+                dialogProps: {channel},
+            }),
+        );
+        trackEvent('ui', 'ui_private_channel_x_button_clicked');
+    };
 
-        this.state = {
-            openUp: false,
-        };
+    const getCloseHandler = () => {
+        if (
+            channel.type === Constants.OPEN_CHANNEL &&
+            channel.name !== Constants.DEFAULT_CHANNEL
+        ) {
+            return handleLeavePublicChannel;
+        } else if (channel.type === Constants.PRIVATE_CHANNEL) {
+            return handleLeavePrivateChannel;
+        }
 
-        this.isLeaving = false;
-    }
+        return null;
+    };
 
-    markAsRead = () => {
-        this.props.actions.markChannelAsRead(this.props.channel.id);
+    const markAsRead = () => {
+        dispatch(markChannelAsRead(channel.id));
         trackEvent('ui', 'ui_sidebar_channel_menu_markAsRead');
-    }
+    };
 
-    favoriteChannel = () => {
-        this.props.actions.favoriteChannel(this.props.channel.id);
+    const favorite = () => {
```

Comment:
`favorite` and `unfavorite` could be consolidated into a single `toggleFavorite` handler